### PR TITLE
GFNet-H-B Model Implementation and Supporting Files (Student ID: S47779628)

### PR DESCRIPTION
**Pull request by Shogo5th**
Hello,

I am submitting this pull request to add the full implementation of the GFNet-H-B model and its supporting files to the correct branch. Below is a summary of the changes:

Changes:
- GFNet Model Implementation (GFNet-H-B):
- A full implementation of the GFNet-H-B architecture was added in modules.py.
- Training Script (train.py): This script initializes the GFNet model and sets up training configurations, including hyperparameters, training, validation, and tuning settings.
- Data Handling (dataset.py): Specifies data loading procedures with necessary preprocessing and data augmentation.
Prediction Script (predict.py): This script loads the trained model with settings for data preprocessing and model configurations. It evaluates the model on test data, calculates accuracy, and plots the training and validation loss from the CSV file generated during training.
Utility Functions (utils.py): This includes utility functions, such as EarlyStopping, to assist with model training.
- README:
  - Updated documentation to explain the usage, setup, and structure of the GFNet implementation. 
  - Added tuning and test results, detailing model performance on hyperparameter tuning and final test accuracy.

Testing:
The model and scripts have been tested on the UQ Rangpur Cluster (for training) and locally (predicting and plotting) to ensure they run as expected and produce the correct results.

Please review the changes and let me know if any adjustments or additional documentation are needed. I am prepared to incorporate any feedback to make sure it matches the project standards.

Thank you for your consideration and review!

Shogo Terashima
Student ID: S47779628

### CODE REVIEW

**Comment by wangzhaomxy** [comment]
**<This is an initial inspection, no action is required at this point.>**

**File Organizing:** Well-organized files.

**Problem Solving:**

- The algorithm solves the problem appropriately. 
- Accuracy in the testing dataset: 0.9862/0.5668.  Student explored 15 trials in the experiment and got 0.5668 testing accuracy based on the original dataset, while after combing the original train/test dataset and resplitting it into a new train/val/test dataset, the testing accuracy can reach 0.9862.

**Model and functions:**

- It correctly uses PyTorch to construct the GFNet models and functions.
- With data augmentation.

**Code design:** Good.

**Code comment and docstring:**

- Perfect code comments 
- Perfect function docstrings

**Difficulty:** Hard.

**Additional Comments:**

- Excellent experiments. 15 trials are explored in the experiment though the best result appears in the re-splitted dataset.
- Reasonable commits.
- Excellent ReadMe design.

**Comment by aniketgupta17** [comment]
Observational Feedback

Pull Request:
Correctly created the Pull request from Topic Recognition Branch .
The pull request shows a clear Description . 

File Organizing: Well-organized files.

Commit Log:
Commit messages could be more  meaningful .
Commits are regularly made.

Documentation:
The README file provides a clear explanation of the algorithm and the problem it solves.
Code comments and docstrings are included .
Proper GitHub markdown formatting is used, with organized headings, lists, and code blocks.
Can add about Future works in  the report

**Comment by gayanku** [comment]
"## Marking
<table><tbody>
  <tr><td colspan='3'><b>Good/OK/Fair Practice (Design/Commenting, TF/Torch Usage)</b></td> </tr>
	<tr><td></td><td>Good design and implementation. </td><td></td></tr>
	<tr><td></td><td>Spacing and comments. </td><td></td></tr>
	<tr><td></td><td>Header blocks. </td><td></td></tr>
	
  <tr><td colspan='3'><b>Recognition Problem</b></td></tr>
	<tr><td></td><td>Good solution to problem. </td><td></td></tr>
	<tr><td></td><td>Driver Script present. </td><td></td></tr>
	<tr><td></td><td>File structure present. </td><td></td></tr>
	<tr><td></td><td>Good Usage & Demo & Visualisation & Data usage. </td><td></td></tr>
	<tr><td></td><td>Module present. </td><td></td></tr>
	<tr><td></td><td>Commenting present. </td><td></td></tr>
	<tr><td></td><td>No Data leakage found. Uses Data Aug.</td><td></td></tr>
	<tr><td></td><td>Difficulty : Hard.  GFNet (Hard Difficulty)</td><td></td></tr>
	
  <tr><td colspan='3'><b>Commit Log</b></td></tr>
	<tr><td></td><td>Good Meaningful commit messages. </td><td></td></tr>
	<tr><td></td><td>Good Progressive commits. </td><td></td></tr>
	
  <tr><td colspan='3'><b>Documentation</b></td></tr>
	<tr><td></td><td>Readme :Good. </td><td></td></tr>
	<tr><td></td><td>Model/technical explanation :Good. </td><td></td></tr>
	<tr><td></td><td>Description and Comments :Good. </td><td></td></tr>
	<tr><td></td><td>Markdown used and PDF submitted. PDF Checked.</td><td></td></tr>
	
  <tr><td colspan='3'><b>Pull Request</b></td></tr>
	<tr><td></td><td>Successful Pull Request (Working Algorithm Delivered on Time in Correct Branch). </td><td></td></tr>
	<tr><td></td><td>No Feedback required. </td><td></td></tr>
	<tr><td></td><td>Request Description is good. </td><td></td></tr>
  <tr><td colspan='2'><b>TOTAL</b></td><td>0</td></tr>
</tbody></table>

Marked as per the due date and changes after which aren't necessarily allowed to contribute to grade for fairness.
_Subject to approval from Shakes_

**Comment by shakes76** [comment]
Great work!
